### ISSUE TITLE
Add a way to parse all numbers as floats in Expression

### ISSUE THREAD

**Issue by me2beats** [godotengine/godot-proposals#5344]
### Describe the project you are working on

A simple calculator

### Describe the problem or limitation you are having in your project

I'm making a simple calculator, just `+-/*.`
The whole expression is visible like `3/2`
And when you press `enter` it is evaluated.
The problem is 3/2 gives 1 because it treats numbers 3 and 2 as integers, and I want to treat all numbers as floats.

So now I should:
- make the user to write 2.0 and 3.0 which is a bad idea or
- replace all ints to floats manually, maybe with Regex <- gonna try this

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I suggest an option to treat all Expression numbers as floats

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Could be a boolean argument like
```
parse(expression: String, input_names: PoolStringArray = PoolStringArray(  ), parse_as_floats = false)
```
false by default

### If this enhancement will not be used often, can it be worked around with a few lines of script?

maybe I can convert all numbers to floats in the string using Regex, but I should also consider cases when for example some numbers are already written in float etc

### Is there a reason why this should be core and not an add-on in the asset library?

I think it would be useful when creating calculators and things like this.
I believe there are other cases when you would like to have this option, even for use in the core

**Comment by Mickeon** [comment]
Could be a flag to expand the concept a bit more. Although, I'm not sure what *else* could go with it.

**Comment by dalexeev** [comment]
You can convert integers to floats using regular expressions, or you can implement your own expression parser. This is not difficult: [dalexeev/gdscript-rpn-calculator](https://github.com/dalexeev/gdscript-rpn-calculator). Also, it should be safer (`VariantParser` doesn't have an allow_objects parameter) and more flexible, since you can control the set of operators and functions available to the user, in `Expression` this set is quite large.

> The problem is 3/2 gives 1 because it treats numbers 3 and 2 as integers

I still think this is a major flaw in GDScript (and the core when it comes to `Expression`), see #1866. But, unfortunately, the vision of the maintainers here is different.

**Comment by dmaz** [comment]
I agree that this is a design flaw in the language... my guess is many of us have run across a bug in our code caused by this... because of this I very rarely even define vars as ints anymore.   

division is and odd duck, I think it should always return float because otherwise are lose precision.  that should always be the conscience decision.  I mean how many times do we actually want to drop the remainder vs keep it.  and of those times wouldn't be more clear to declare that?

**Comment by Mickeon** [comment]
Most languages I know return a `float` from a division only when one of the two operands is a `float`. If both are `integers`, the result is an `integer`. I find this to be the most expected outcome. With this behavior, to do as current GDScript does, one would simply need to **explicitly** convert one or both operands to `float` (which is similar to what GDScript does behind the scenes, anyhow).

**Comment by dmaz** [comment]
actually... https://github.com/godotengine/godot-proposals/issues/1866 was a good proposal just the wrong way around... maybe we should have // be a never lose precision division?

**Comment by dalexeev** [comment]
This proposal is not about the division operator. Please discuss this at #1866.

**Comment by dmaz** [comment]
I disagree... the op described the limitation as the default division operator and thus gave a proposal.  changing the division operator is clearly and alternate solution to his issue.   and 1866 is closed.  but fine.
